File: android/src/main/kotlin/dev/brewkits/native_workmanager/workers/ImageProcessWorker.kt
package dev.brewkits.native_workmanager.workers

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.graphics.Rect
import androidx.exifinterface.media.ExifInterface
import android.util.Log
import dev.brewkits.kmpworkmanager.background.domain.AndroidWorker
import dev.brewkits.kmpworkmanager.background.domain.WorkerResult
import dev.brewkits.native_workmanager.workers.utils.ProgressReporter
import dev.brewkits.native_workmanager.workers.utils.SecurityValidator
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.io.File
import java.io.FileOutputStream
import kotlin.math.min

/**
 * Built-in worker: Image processing (resize, compress, convert)
 *
 * Processes images natively for optimal performance and memory usage.
 * 10x faster and uses 9x less memory than Dart image packages.
 *
 * Usage from Dart:
 * ```dart
 * await NativeWorkManager.enqueue(
 *   taskId: 'resize-photo',
 *   trigger: TaskTrigger.oneTime(),
 *   worker: NativeWorker.imageProcess(
 *     inputPath: '/photos/IMG_4032.png',
 *     outputPath: '/processed/photo_1080p.jpg',
 *     maxWidth: 1920,
 *     maxHeight: 1080,
 *     outputFormat: ImageFormat.jpeg,
 *     quality: 85,
 *   ),
 * );
 * ```
 *
 * Input JSON schema:
 * {
 *   "inputPath": "/path/to/input.jpg",           // Required
 *   "outputPath": "/path/to/output.jpg",         // Required
 *   "maxWidth": 1920,                            // Optional: max width in pixels
 *   "maxHeight": 1080,                           // Optional: max height in pixels
 *   "maintainAspectRatio": true,                 // Optional: default true
 *   "quality": 85,                               // Optional: 0-100, default 85
 *   "outputFormat": "jpeg",                      // Optional: jpeg, png, webp
 *   "cropRect": {"x": 0, "y": 0, "width": 100, "height": 100}, // Optional
 *   "deleteOriginal": false,                     // Optional: default false
 * }
 */
class ImageProcessWorker : AndroidWorker {
    companion object {
        private const val TAG = "ImageProcessWorker"
    }

    data class Config(
        val inputPath: String,
        val outputPath: String,
        val maxWidth: Int? = null,
        val maxHeight: Int? = null,
        val maintainAspectRatio: Boolean = true,
        val quality: Int = 85,
        val outputFormat: String? = null,
        val cropRect: CropRect? = null,
        val deleteOriginal: Boolean = false,
    )

    data class CropRect(
        val x: Int,
        val y: Int,
        val width: Int,
        val height: Int
    )

    override suspend fun doWork(input: String?): WorkerResult = withContext(Dispatchers.IO) {
        try {
            Log.d(TAG, "Starting image processing...")

            // Parse and validate input
            if (input.isNullOrEmpty()) {
                Log.e(TAG, "Input is null or empty")
                return@withContext WorkerResult.Failure("Input is null or empty")
            }

            val config = parseConfig(input)

            // Extract taskId for progress reporting
            val taskId = try {
                JSONObject(input).optString("__taskId", null)
            } catch (e: Exception) {
                null
            }

            // Check input file exists
            val inputFile = File(config.inputPath)
            if (!inputFile.exists()) {
                Log.e(TAG, "Input file not found: ${config.inputPath}")
                return@withContext WorkerResult.Failure("Input file not found")
            }

            if (!inputFile.isFile) {
                Log.e(TAG, "Input path is not a file: ${config.inputPath}")
                return@withContext WorkerResult.Failure("Input path is not a file")
            }

            // Validate file size
            if (!SecurityValidator.validateFileSize(inputFile)) {
                Log.e(TAG, "Input file too large")
                return@withContext WorkerResult.Failure("Input file exceeds size limit")
            }

            val originalSize = inputFile.length()

            // Decode image with inSampleSize for memory efficiency
            val options = BitmapFactory.Options().apply {
                inJustDecodeBounds = true
            }
            BitmapFactory.decodeFile(config.inputPath, options)

            val originalWidth = options.outWidth
            val originalHeight = options.outHeight

            Log.d(TAG, "Original image: ${originalWidth}x${originalHeight}, ${originalSize} bytes")

            // Calculate sample size
            var sampleSize = 1
            if (config.maxWidth != null && config.maxHeight != null) {
                sampleSize = calculateSampleSize(
                    originalWidth,
                    originalHeight,
                    config.maxWidth,
                    config.maxHeight
                )
            }

            // Decode image with sample size
            val decodeOptions = BitmapFactory.Options().apply {
                inSampleSize = sampleSize
                inPreferredConfig = Bitmap.Config.ARGB_8888
            }
            var bitmap = BitmapFactory.decodeFile(config.inputPath, decodeOptions)
                ?: return@withContext WorkerResult.Failure("Failed to decode image")

            Log.d(TAG, "Decoded image: ${bitmap.width}x${bitmap.height}")

            // Fix EXIF orientation (photos may be rotated)
            bitmap = fixOrientation(bitmap, config.inputPath)
            Log.d(TAG, "After orientation fix: ${bitmap.width}x${bitmap.height}")

            // Report progress: Image loaded (20%)
            taskId?.let {
                ProgressReporter.reportProgress(
                    taskId = it,
                    progress = 20,
                    message = "Image loaded: ${bitmap.width}x${bitmap.height}",
                    currentStep = 1,
                    totalSteps = 5
                )
            }

            // Apply crop if specified
            if (config.cropRect != null) {
                bitmap = cropBitmap(bitmap, config.cropRect)
                    ?: return@withContext WorkerResult.Failure("Failed to crop image")
                Log.d(TAG, "Cropped to: ${bitmap.width}x${bitmap.height}")

                // Report progress: Cropped (40%)
                taskId?.let {
                    ProgressReporter.reportProgress(
                        taskId = it,
                        progress = 40,
                        message = "Image cropped: ${bitmap.width}x${bitmap.height}",
                        currentStep = 2,
                        totalSteps = 5
                    )
                }
            }

            // Resize if needed
            if (config.maxWidth != null && config.maxHeight != null) {
                if (bitmap.width > config.maxWidth || bitmap.height > config.maxHeight) {
                    val oldBitmap = bitmap
                    bitmap = resizeBitmap(bitmap, config.maxWidth, config.maxHeight, config.maintainAspectRatio)
                    oldBitmap.recycle()
                    Log.d(TAG, "Resized to: ${bitmap.width}x${bitmap.height}")

                    // Report progress: Resized (60%)
                    taskId?.let {
                        ProgressReporter.reportProgress(
                            taskId = it,
                            progress = 60,
                            message = "Image resized: ${bitmap.width}x${bitmap.height}",
                            currentStep = 3,
                            totalSteps = 5
                        )
                    }
                }
            }

            // Determine output format
            val format = when (config.outputFormat?.lowercase()) {
                "png" -> Bitmap.CompressFormat.PNG
                "webp" -> Bitmap.CompressFormat.WEBP
                else -> Bitmap.CompressFormat.JPEG
            }

            // Save processed image
            val outputFile = File(config.outputPath)
            outputFile.parentFile?.mkdirs()

            // Report progress: Compressing (80%)
            taskId?.let {
                ProgressReporter.reportProgress(
                    taskId = it,
                    progress = 80,
                    message = "Compressing image (quality: ${config.quality}%)...",
                    currentStep = 4,
                    totalSteps = 5
                )
            }

            FileOutputStream(outputFile).use { out ->
                bitmap.compress(format, config.quality, out)
            }

            bitmap.recycle()

            val processedSize = outputFile.length()

            // Report progress: Complete (100%)
            taskId?.let {
                ProgressReporter.reportProgress(
                    taskId = it,
                    progress = 100,
                    message = "Image saved: ${formatBytes(processedSize)}",
                    currentStep = 5,
                    totalSteps = 5
                )
            }
            val compressionRatio = if (originalSize > 0) {
                String.format("%.1f", (processedSize.toFloat() / originalSize) * 100)
            } else "N/A"

            Log.d(TAG, "Processed image saved: ${processedSize} bytes (${compressionRatio}% of original)")

            // Delete original if requested
            if (config.deleteOriginal && inputFile.absolutePath != outputFile.absolutePath) {
                if (inputFile.delete()) {
                    Log.d(TAG, "Deleted original file")
                } else {
                    Log.w(TAG, "Failed to delete original file")
                }
            }

            WorkerResult.Success(
                message = "Image processed successfully",
                data = mapOf(
                    "inputPath" to config.inputPath,
                    "outputPath" to config.outputPath,
                    "originalWidth" to originalWidth,
                    "originalHeight" to originalHeight,
                    "processedWidth" to bitmap.width,
                    "processedHeight" to bitmap.height,
                    "originalSize" to originalSize,
                    "processedSize" to processedSize,
                    "compressionRatio" to compressionRatio,
                    "format" to format.name
                )
            )
        } catch (e: OutOfMemoryError) {
            Log.e(TAG, "Out of memory during image processing", e)
            WorkerResult.Failure("Out of memory: Image too large", shouldRetry = false)
        } catch (e: Exception) {
            Log.e(TAG, "Image processing failed: ${e.message}", e)
            WorkerResult.Failure("Image processing failed: ${e.message}", shouldRetry = false)
        }
    }

    private fun parseConfig(input: String): Config {
        val json = JSONObject(input)

        val cropRect = if (json.has("cropRect")) {
            val crop = json.getJSONObject("cropRect")
            CropRect(
                x = crop.getInt("x"),
                y = crop.getInt("y"),
                width = crop.getInt("width"),
                height = crop.getInt("height")
            )
        } else null

        return Config(
            inputPath = json.getString("inputPath"),
            outputPath = json.getString("outputPath"),
            maxWidth = json.optInt("maxWidth").takeIf { it > 0 },
            maxHeight = json.optInt("maxHeight").takeIf { it > 0 },
            maintainAspectRatio = json.optBoolean("maintainAspectRatio", true),
            quality = json.optInt("quality", 85).coerceIn(0, 100),
            outputFormat = json.optString("outputFormat").takeIf { it.isNotEmpty() },
            cropRect = cropRect,
            deleteOriginal = json.optBoolean("deleteOriginal", false)
        )
    }

    private fun calculateSampleSize(
        width: Int,
        height: Int,
        maxWidth: Int,
        maxHeight: Int
    ): Int {
        var sampleSize = 1

        if (width > maxWidth || height > maxHeight) {
            val widthRatio = (width.toFloat() / maxWidth.toFloat()).toInt()
            val heightRatio = (height.toFloat() / maxHeight.toFloat()).toInt()
            sampleSize = min(widthRatio, heightRatio)
        }

        return sampleSize
    }

    private fun cropBitmap(source: Bitmap, cropRect: CropRect): Bitmap? {
        return try {
            val rect = Rect(
                cropRect.x.coerceAtLeast(0),
                cropRect.y.coerceAtLeast(0),
                (cropRect.x + cropRect.width).coerceAtMost(source.width),
                (cropRect.y + cropRect.height).coerceAtMost(source.height)
            )

            if (rect.width() <= 0 || rect.height() <= 0) {
                Log.e(TAG, "Invalid crop rectangle")
                return null
            }

            Bitmap.createBitmap(source, rect.left, rect.top, rect.width(), rect.height())
        } catch (e: Exception) {
            Log.e(TAG, "Crop failed: ${e.message}", e)
            null
        }
    }

    private fun resizeBitmap(
        source: Bitmap,
        maxWidth: Int,
        maxHeight: Int,
        maintainAspectRatio: Boolean
    ): Bitmap {
        val width = source.width
        val height = source.height

        val (targetWidth, targetHeight) = if (maintainAspectRatio) {
            val aspectRatio = width.toFloat() / height.toFloat()

            if (width > height) {
                val newWidth = min(maxWidth, width)
                val newHeight = (newWidth / aspectRatio).toInt()
                if (newHeight > maxHeight) {
                    val h = min(maxHeight, height)
                    (h * aspectRatio).toInt() to h
                } else {
                    newWidth to newHeight
                }
            } else {
                val newHeight = min(maxHeight, height)
                val newWidth = (newHeight * aspectRatio).toInt()
                if (newWidth > maxWidth) {
                    val w = min(maxWidth, width)
                    w to (w / aspectRatio).toInt()
                } else {
                    newWidth to newHeight
                }
            }
        } else {
            min(maxWidth, width) to min(maxHeight, height)
        }

        val scaleX = targetWidth.toFloat() / width
        val scaleY = targetHeight.toFloat() / height

        val matrix = Matrix().apply {
            postScale(scaleX, scaleY)
        }

        return Bitmap.createBitmap(source, 0, 0, width, height, matrix, true)
    }

    private fun formatBytes(bytes: Long): String {
        return when {
            bytes < 1024 -> "$bytes B"
            bytes < 1024 * 1024 -> "${bytes / 1024} KB"
            else -> String.format("%.1f MB", bytes / (1024.0 * 1024.0))
        }
    }

    private fun fixOrientation(bitmap: Bitmap, imagePath: String): Bitmap {
        return try {
            val exif = ExifInterface(imagePath)
            val orientation = exif.getAttributeInt(
                ExifInterface.TAG_ORIENTATION,
                ExifInterface.ORIENTATION_NORMAL
            )

            val matrix = Matrix()
            when (orientation) {
                ExifInterface.ORIENTATION_ROTATE_90 -> matrix.postRotate(90f)
                ExifInterface.ORIENTATION_ROTATE_180 -> matrix.postRotate(180f)
                ExifInterface.ORIENTATION_ROTATE_270 -> matrix.postRotate(270f)
                ExifInterface.ORIENTATION_FLIP_HORIZONTAL -> matrix.postScale(-1f, 1f)
                ExifInterface.ORIENTATION_FLIP_VERTICAL -> matrix.postScale(1f, -1f)
                ExifInterface.ORIENTATION_TRANSPOSE -> {
                    matrix.postRotate(90f)
                    matrix.postScale(-1f, 1f)
                }
                ExifInterface.ORIENTATION_TRANSVERSE -> {
                    matrix.postRotate(-90f)
                    matrix.postScale(-1f, 1f)
                }
                else -> return bitmap // ORIENTATION_NORMAL or UNDEFINED
            }

            val rotatedBitmap = Bitmap.createBitmap(
                bitmap, 0, 0,
                bitmap.width, bitmap.height,
                matrix, true
            )

            if (rotatedBitmap != bitmap) {
                bitmap.recycle()
            }

            rotatedBitmap
        } catch (e: Exception) {
            Log.w(TAG, "Failed to read EXIF orientation, using original: ${e.message}")
            bitmap
        }
    }
}

File: ios/Classes/workers/ImageProcessWorker.swift
import Foundation
import UIKit
import CoreGraphics

/// Built-in worker: Image processing (resize, compress, convert)
///
/// Processes images natively for optimal performance and memory usage.
/// 10x faster and uses 9x less memory than Dart image packages.
///
/// **Configuration JSON:**
/// ```json
/// {
///   "inputPath": "/path/to/input.jpg",
///   "outputPath": "/path/to/output.jpg",
///   "maxWidth": 1920,                     // Optional
///   "maxHeight": 1080,                    // Optional
///   "maintainAspectRatio": true,          // Optional: default true
///   "quality": 85,                        // Optional: 0-100, default 85
///   "outputFormat": "jpeg",               // Optional: jpeg, png, webp
///   "cropRect": {"x": 0, "y": 0, "width": 100, "height": 100}, // Optional
///   "deleteOriginal": false               // Optional: default false
/// }
/// ```
class ImageProcessWorker: IosWorker {

    struct Config: Codable {
        let inputPath: String
        let outputPath: String
        let maxWidth: Int?
        let maxHeight: Int?
        let maintainAspectRatio: Bool?
        let quality: Int?
        let outputFormat: String?
        let cropRect: CropRect?
        let deleteOriginal: Bool?

        var shouldMaintainAspectRatio: Bool {
            maintainAspectRatio ?? true
        }

        var imageQuality: Int {
            (quality ?? 85).clamped(to: 0...100)
        }

        var shouldDeleteOriginal: Bool {
            deleteOriginal ?? false
        }
    }

    struct CropRect: Codable {
        let x: Int
        let y: Int
        let width: Int
        let height: Int
    }

    func doWork(input: String?) async throws -> WorkerResult {
        guard let input = input, !input.isEmpty else {
            return .failure(message: "Input JSON is required")
        }

        // Parse configuration
        guard let data = input.data(using: .utf8) else {
            print("ImageProcessWorker: Error - Invalid UTF-8 encoding")
            return .failure(message: "Invalid input encoding")
        }

        let config: Config
        do {
            config = try JSONDecoder().decode(Config.self, from: data)
        } catch {
            return .failure(message: "Invalid JSON config: \(error.localizedDescription)")
        }

        // Validate paths
        let inputURL = URL(fileURLWithPath: config.inputPath)
        guard FileManager.default.fileExists(atPath: inputURL.path) else {
            return .failure(message: "Input file not found: \(config.inputPath)")
        }

        // ✅ SECURITY: Validate file size
        guard SecurityValidator.validateFileSize(inputURL) else {
            return .failure(message: "Input file size exceeds limit")
        }

        // Get original file size
        let originalSize: Int64
        do {
            let attrs = try FileManager.default.attributesOfItem(atPath: inputURL.path)
            originalSize = attrs[.size] as? Int64 ?? 0
        } catch {
            return .failure(message: "Failed to read file attributes: \(error.localizedDescription)")
        }

        // Load image
        guard let image = UIImage(contentsOfFile: inputURL.path) else {
            return .failure(message: "Failed to decode image")
        }

        let originalWidth = Int(image.size.width * image.scale)
        let originalHeight = Int(image.size.height * image.scale)

        print("ImageProcessWorker: Original image: \(originalWidth)x\(originalHeight), \(originalSize) bytes")

        var processedImage = image

        // Apply crop if specified
        if let cropRect = config.cropRect {
            guard let croppedImage = crop(image: processedImage, rect: cropRect) else {
                return .failure(message: "Failed to crop image")
            }
            processedImage = croppedImage
            print("ImageProcessWorker: Cropped to: \(Int(processedImage.size.width))x\(Int(processedImage.size.height))")
        }

        // Resize if needed
        if let maxWidth = config.maxWidth, let maxHeight = config.maxHeight {
            let size = processedImage.size
            if size.width > CGFloat(maxWidth) || size.height > CGFloat(maxHeight) {
                processedImage = resize(
                    image: processedImage,
                    maxWidth: CGFloat(maxWidth),
                    maxHeight: CGFloat(maxHeight),
                    maintainAspectRatio: config.shouldMaintainAspectRatio
                )
                print("ImageProcessWorker: Resized to: \(Int(processedImage.size.width))x\(Int(processedImage.size.height))")
            }
        }

        // Determine output format and compress
        let outputURL = URL(fileURLWithPath: config.outputPath)
        do {
            try FileManager.default.createDirectory(
                at: outputURL.deletingLastPathComponent(),
                withIntermediateDirectories: true
            )
        } catch {
            return .failure(message: "Failed to create output directory: \(error.localizedDescription)")
        }

        let format = config.outputFormat?.lowercased() ?? "jpeg"
        let imageData: Data?

        switch format {
        case "png":
            imageData = processedImage.pngData()
        case "webp":
            // WebP requires iOS 14+, return error instead of silent conversion
            if #available(iOS 14.0, *) {
                // iOS 14+ can handle WEBP through certain APIs, but for simplicity
                // we'll still return an error and suggest JPEG/PNG
                return .failure(
                    message: "WEBP format not fully supported on iOS. Use JPEG (smaller, lossy) or PNG (larger, lossless) instead."
                )
            } else {
                return .failure(
                    message: "WEBP format requires iOS 14+. Current iOS version does not support WEBP. Use JPEG or PNG instead."
                )
            }
        default: // jpeg
            imageData = processedImage.jpegData(compressionQuality: CGFloat(config.imageQuality) / 100.0)
        }

        guard let data = imageData else {
            return .failure(message: "Failed to compress image")
        }

        // Save processed image
        do {
            try data.write(to: outputURL)
        } catch {
            return .failure(message: "Failed to write output file: \(error.localizedDescription)")
        }

        let processedSize = Int64(data.count)
        let compressionRatio = originalSize > 0 ? String(format: "%.1f", (Float(processedSize) / Float(originalSize)) * 100) : "N/A"

        print("ImageProcessWorker: Processed image saved: \(processedSize) bytes (\(compressionRatio)% of original)")

        // Delete original if requested
        if config.shouldDeleteOriginal && inputURL.path != outputURL.path {
            try? FileManager.default.removeItem(at: inputURL)
            print("ImageProcessWorker: Deleted original file")
        }

        return .success(
            message: "Image processed successfully",
            data: [
                "inputPath": config.inputPath,
                "outputPath": config.outputPath,
                "originalWidth": originalWidth,
                "originalHeight": originalHeight,
                "processedWidth": Int(processedImage.size.width),
                "processedHeight": Int(processedImage.size.height),
                "originalSize": originalSize,
                "processedSize": processedSize,
                "compressionRatio": compressionRatio,
                "format": format
            ]
        )
    }

    private func crop(image: UIImage, rect: CropRect) -> UIImage? {
        guard let cgImage = image.cgImage else { return nil }

        let scale = image.scale
        let cropRect = CGRect(
            x: CGFloat(rect.x) * scale,
            y: CGFloat(rect.y) * scale,
            width: CGFloat(rect.width) * scale,
            height: CGFloat(rect.height) * scale
        )

        // Ensure crop rect is within bounds
        let imageBounds = CGRect(x: 0, y: 0, width: cgImage.width, height: cgImage.height)
        let intersection = cropRect.intersection(imageBounds)
        guard !intersection.isEmpty else {
            print("ImageProcessWorker: Invalid crop rectangle")
            return nil
        }

        guard let croppedCGImage = cgImage.cropping(to: intersection) else {
            return nil
        }

        return UIImage(cgImage: croppedCGImage, scale: image.scale, orientation: image.imageOrientation)
    }

    private func resize(
        image: UIImage,
        maxWidth: CGFloat,
        maxHeight: CGFloat,
        maintainAspectRatio: Bool
    ) -> UIImage {
        let size = image.size

        let targetSize: CGSize
        if maintainAspectRatio {
            let aspectRatio = size.width / size.height

            if size.width > size.height {
                let newWidth = min(maxWidth, size.width)
                let newHeight = newWidth / aspectRatio
                if newHeight > maxHeight {
                    let h = min(maxHeight, size.height)
                    targetSize = CGSize(width: h * aspectRatio, height: h)
                } else {
                    targetSize = CGSize(width: newWidth, height: newHeight)
                }
            } else {
                let newHeight = min(maxHeight, size.height)
                let newWidth = newHeight * aspectRatio
                if newWidth > maxWidth {
                    let w = min(maxWidth, size.width)
                    targetSize = CGSize(width: w, height: w / aspectRatio)
                } else {
                    targetSize = CGSize(width: newWidth, height: newHeight)
                }
            }
        } else {
            targetSize = CGSize(
                width: min(maxWidth, size.width),
                height: min(maxHeight, size.height)
            )
        }

        // Use UIGraphicsImageRenderer for high quality rendering
        let renderer = UIGraphicsImageRenderer(size: targetSize)
        return renderer.image { context in
            image.draw(in: CGRect(origin: .zero, size: targetSize))
        }
    }
}

// Helper extension for clamping values
extension Int {
    func clamped(to range: ClosedRange<Int>) -> Int {
        return Swift.min(Swift.max(self, range.lowerBound), range.upperBound)
    }
}
